src/index.js

```diff
@@ -19,41 +19,96 @@ const debug = (function () {
 })()
 
 class Masq {
-  /**
-   * constructor
-   * @param {string} appName - The application name
-   */
   constructor (appName, appDescription, appImageURL, options = {}) {
-    this._reset()
-
     this.eventTarget = document.createElement('MasqLib')
 
     this.appName = appName
     this.appDescription = appDescription
     this.appImageURL = appImageURL
 
-    this.userId = null
-    this.dataEncryptionKey = null
-    this.userAppDb = null
-
-    this._loadSessionInfo()
-
-    // login state data
-    this.loginChannel = null
-    this.loginKey = null
-    this.loginUrl = null
-
     // override config with constructor options
     this.config = {
       hubUrls: options.hubUrls ? options.hubUrls : jsonConfig.hubUrls,
       masqAppBaseUrl: options.masqAppBaseUrl ? options.masqAppBaseUrl : jsonConfig.masqAppBaseUrl,
       swarmConfig: options.swarmConfig ? options.swarmConfig : jsonConfig.swarmConfig
     }
 
+    // init state
+    this.setState('notLogged')
+
+    // init state through async function
+    // this.init should be awaited to make sure state is initialized
+    this.initPromise = this._init()
+
+    // login state data
+    this.stayConnected = false
+    this.loginChannel = null
+    this.loginKey = null
+    this.loginSw = null
+    this.loginPeer = null
+
+    // reset replication variables
+    this.userAppRepSW = null
+    this.userAppRepHub = null
+
+    // setup listeners
     this._listenForLoginOrSignout()
     this._listenForFocus()
   }
 
+  setState (newState) {
+    const currState = this.state
+    switch (newState) {
+      case 'notLogged':
+        break
+      case 'userAppDbCreated':
+        break
+      case 'accessMaterialReceived':
+        break
+      case 'registerNeeded':
+        break
+      case 'authorized':
+        break
+      case 'loggingIn':
+        break
+      case 'logged':
+        if (currState !== 'notLogged' &&
+            currState !== 'authorized' &&
+            currState !== 'userAppDbCreated' &&
+            currState !== 'replicating') {
+          throw new MasqError(
+            MasqError.INVALID_STATE_TRANSITION,
+            'state transition invalid: ' + currState + ' -> ' + newState)
+        }
+        // dispatch event logged_in if the transition is not from replicating
+        if (currState !== 'replicating') {
+          debug('[masq.setState] dispatched event: logged_in')
+          this.eventTarget.dispatchEvent(new Event('logged_in'))
+        }
+        break
+      case 'replicating':
+        break
+      default:
+        throw new MasqError()
+    }
+    debug('[masq.setState] changing state : ' + currState + ' -> ' + newState)
+    this.state = newState
+  }
+
+  async init () {
```

Comment:
as we already dispatch multiple event like `logged_in', I find it would be more consistent to ask the user to wait on a `ready` event, what do you think ?

Reply:
I don't like the `ready` event in general, the other events do not necessarily occur at the begining while the `ready` event will necessarily occur right after instantiation. 
If we expose a ready event I think in most case we will await on it right after instantiation.
I prefer to expose a promise that will be resolved when masq is ready.